### PR TITLE
Outcome refactor and leader commit

### PR DESCRIPTION
**Pull request by martinfurmanski**
Refactor outcome to encapsulate boiler plate and clarify logic in roles and only consider leader commit updates if history matches.

### CODE REVIEW

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/outcome/Outcome.java [hunk 1]
```diff
@@ -20,54 +20,144 @@
...
+    public MEMBER votedFor;
+    public boolean renewElectionTimeout;
+
+    /* Candidate */
+    public HashSet<MEMBER> votesForMe;
+    public long lastLogIndexBeforeWeBecameLeader;
+
+    /* Leader */
+    public FollowerStates<MEMBER> followerStates;
+    public ArrayList<ShipCommand> shipCommands = new ArrayList<>();
```
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/outcome/Outcome.java:64]: Having non final public fields sounds pretty strange. Why did you remove `final`s?

Comment (lutovich) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/outcome/Outcome.java:64]: @martinfurmanski ping. Please answer the question. This is a valid point/concern.

Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/outcome/Outcome.java:64]: @davidegrohmann The finals were removed so that the fields could be modified. What is your actual concern?

Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/outcome/Outcome.java:64]: @martinfurmanski Having public non-final fields is a bad design in principle imho. Here it is even more confusing what you are trying to achieve since you have non-final public fields and setters.

Comment (chrisvest) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/outcome/Outcome.java:64]: Removing the finals also changes the concurrency semantics, in case instances of this class are shared between multiple threads.

**Comment by martinfurmanski** [comment]
@chrisvest Sure, but that was not the reason for it being immutable before. Net change zero.

**Comment by martinfurmanski** [comment]
@davidegrohmann Outcome was just a bag of stuff and still is. I didn't design the encapsulation and it wasn't really within the scope of this PR to change it so that the implementation got hidden underneath an interface. There isn't much of an implementation, although a little bit more now than before. The reason for this change was mainly to make the follower/candidate/leader logic code easier to follow and read, since it was causing trouble previously when it wasn't clear when the actual outcome was being updated. I don't disagree that public non-final fields in general is a bad design (not as much in this case though). I will take a look at it anyway though, but it will need changes to hundreds of lines of code in tests that already depend on this level of encapsulation, so it will be part of another refactoring in that case.
